deals/service.go

```diff
@@ -137,3 +137,34 @@ func (s *Service) Watch(req *pb.WatchRequest, srv pb.API_WatchServer) error {
 	}
 	return nil
 }
+
+// Retrieve calls deals.Retreive
+func (s *Service) Retrieve(req *pb.RetrieveRequest, srv pb.API_RetrieveServer) error {
+	cid, err := cid.Parse(req.GetCid())
+	if err != nil {
+		return err
+	}
+
+	reader, err := s.Module.Retrieve(srv.Context(), req.GetAddress(), cid)
+	defer reader.Close()
+	if err != nil {
+		return err
+	}
+
+	buffer := make([]byte, 1024*32) // 32KB
+	for {
+		bytesRead, err := reader.Read(buffer)
+		if err != nil && err != io.EOF {
+			return nil
+		}
+		sendErr := srv.Send(&pb.RetrieveReply{Chunk: buffer[:bytesRead]})
+		if sendErr != nil {
+			return sendErr
+		}
+		if err == io.EOF {
+			break
+		}
+	}
```

Comment:
Hmmm I suppose we need to react to the `srv.Context` being canceled, and in that case stop reading the `reader` and return?

Reply:
Can apply any feedback about the best way to do that in the client here as well

Reply:
Have an idea here, about to push and will get feedback